### PR TITLE
Created form fields tests.

### PR DESCRIPTION
**Pull request by rralys**
Created test cases for the form fields.

### CODE REVIEW

File: jdi-light-angular/src/main/java/com/epam/jdi/light/angular/elements/complex/FormFields.java [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+import com.epam.jdi.light.elements.complex.*;
+import com.epam.jdi.light.logger.*;
+import com.epam.jdi.tools.map.*;
+import org.openqa.selenium.*;
+
+import java.util.*;
+
+import static com.epam.jdi.light.common.TextTypes.*;
+import static com.epam.jdi.light.driver.WebDriverFactory.*;
+
```
Comment (AlexSofft) [inline jdi-light-angular/src/main/java/com/epam/jdi/light/angular/elements/complex/FormFields.java:17]: /**
 * To see an example of FormField web element please visit https://material.angular.io/components/form-field/overview#form-field-appearance-variants
 */

add comment before class
2.  If it's not critical, change the name of the element class for FormField (single)

**Comment by mf-code** [comment]
[6127] - V2YpaJoWVDZ4

**Comment by mf-code** [comment]
[6127] - 2krj01OGLhr4

File: jdi-light-angular-tests/src/test/java/io/github/epam/angular/tests/unit/FormFieldsUnitTests.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+import io.github.epam.*;
+import org.testng.annotations.*;
+
+import static io.github.com.StaticSite.*;
+import static io.github.com.pages.AngularPage.*;
+import static io.github.epam.site.steps.States.*;
+import static org.testng.Assert.*;
+
+public class FormFieldsUnitTests extends TestsInit {
+    @BeforeMethod
```
Comment (AlexSofft) [inline jdi-light-angular-tests/src/test/java/io/github/epam/angular/tests/unit/FormFieldsUnitTests.java:12]: @BeforeMethod(alwaysRun = true)

**Comment by mf-code** [comment]
[6127] - iHKrexcqhKgh

**Comment by mf-code** [comment]
[6127] - jY9ZAISvZcpl

File: jdi-light-angular-tests/src/test/java/io/github/epam/angular/tests/elements/complex/FormFieldsTests.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+import org.testng.annotations.*;
+
+import static io.github.com.StaticSite.*;
+import static io.github.com.pages.AngularPage.*;
+import static io.github.epam.site.steps.States.*;
+import static org.hamcrest.Matchers.*;
+
+public class FormFieldsTests extends TestsInit {
+
+    @BeforeMethod
```
Comment (AlexSofft) [inline jdi-light-angular-tests/src/test/java/io/github/epam/angular/tests/elements/complex/FormFieldsTests.java:13]: @BeforeMethod(alwaysRun = true)

**Review by AlexSofft — CHANGES_REQUESTED** [review]

**Comment by mf-code** [comment]
[6127] - 6fFomCbC43mB

**Comment by mf-code** [comment]
[6127] - EuK4KrX7hEmy

**Comment by mf-code** [comment]
[6127] - lrLHlTBbb8QB

**Comment by mf-code** [comment]
[6127] - lv6YelcvJMoh

**Comment by mf-code** [comment]
[6127] - y4b7IXWGS9og

**Comment by mf-code** [comment]
[6127] - KYP2u2FdSbzX

**Comment by mf-code** [comment]
[6127] - aX0IlrhIV7Hl

**Comment by mf-code** [comment]
[6127] - tAh7RFfmPDpI

**Comment by mf-code** [comment]
[6127] - PG5rIgMbU934

**Comment by mf-code** [comment]
[6127] - SNZg7mL1fR2j

**Comment by mf-code** [comment]
[6127] - 5c5Iej4bcVjV

**Comment by mf-code** [comment]
[6127] - Lu0wnxbWjHEl

**Comment by mf-code** [comment]
[6127] - qGmCkfifrrPI

**Comment by mf-code** [comment]
[6127] - GRdZSgVs9S1b

**Comment by mf-code** [comment]
[6127] - openjdk8 - Allure report on Netlify: https://5f63250d2bf49639a886b9d5--musing-swartz-6fc060.netlify.app

**Comment by mf-code** [comment]
[6127] - openjdk10 - Allure report on Netlify: https://5f6325ff085c9d44df8d6ce7--musing-swartz-6fc060.netlify.app

**Comment by mf-code** [comment]
[6130] - hvmwPjlh3TDp

**Comment by mf-code** [comment]
[6130] - yuFlXb5uKE0g

**Comment by mf-code** [comment]
[6130] - s3U2uXvsALw5

**Comment by mf-code** [comment]
[6130] - GJPByBTxHh0w

**Comment by mf-code** [comment]
[6130] - 3QuhkzqwKZkj

**Comment by mf-code** [comment]
[6130] - qt4aT7hy1Mvi

**Comment by mf-code** [comment]
[6130] - 8zy2T0N1RsK8

**Comment by mf-code** [comment]
[6130] - iUKP9LTsONY3

**Comment by mf-code** [comment]
[6130] - YXZXhLV1Deiv

**Comment by mf-code** [comment]
[6130] - AnrHbfGotlfl

**Comment by mf-code** [comment]
[6130] - AILbQcMGMWzk

**Comment by mf-code** [comment]
[6130] - GcGnKJbArwgK

**Comment by mf-code** [comment]
[6130] - vLrB5Mx2zHaL

**Comment by mf-code** [comment]
[6130] - rt5eix3tCt6S

**Comment by mf-code** [comment]
[6130] - 7kXR6BxiVRkC

**Comment by mf-code** [comment]
[6130] - GKjRMQp59YJP

**Comment by mf-code** [comment]
[6130] - gbDGYcHbtiTa

**Comment by mf-code** [comment]
[6130] - Evnn6RO54ai0

**Comment by mf-code** [comment]
[6130] - openjdk8 - Allure report on Netlify: https://5f6341c39b99d3b26803d3a5--musing-swartz-6fc060.netlify.app

**Comment by mf-code** [comment]
[6130] - openjdk10 - Allure report on Netlify: https://5f634291b0ade28b08d2205a--musing-swartz-6fc060.netlify.app

File: jdi-light-angular-tests/src/main/java/io/github/com/pages/sections/FormFieldsSection.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+    @UI("#prefix-suffix-form-field-password")
+    public FormField formFieldPrefixSuffix;
+
+    @UI("#modified-layout-form-field-color")
+    public FormField formFieldThemes;
+
+    @UI("#form-field-phone-number-id")
+    public FormField fromFieldsCustomControl;
+
```
Comment (gbuh) [inline jdi-light-angular-tests/src/main/java/io/github/com/pages/sections/FormFieldsSection.java:30]: Please remove empty line

File: jdi-light-angular-tests/src/main/java/io/github/com/pages/sections/FormFieldsSection.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
+package io.github.com.pages.sections;
+
+import com.epam.jdi.light.angular.elements.complex.*;
+import com.epam.jdi.light.elements.pageobjects.annotations.locators.*;
+
+public class FormFieldsSection {
+    @UI("#simple-form-field-input")
```
Comment (gbuh) [inline jdi-light-angular-tests/src/main/java/io/github/com/pages/sections/FormFieldsSection.java:7]: Why aren't you using smart locators. It's simple and also reduces the number of lines of code.

File: jdi-light-angular-tests/src/main/java/io/github/com/pages/sections/FormFieldsSection.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    public FormField simpleFormField;
+
+    @UI("#form-field-with-label-input-simple-placeholder")
```
Comment (gbuh) [inline jdi-light-angular-tests/src/main/java/io/github/com/pages/sections/FormFieldsSection.java:10]: If we create the id="form-field-with-label", the @UI annotation is not needed.

**Review by AlexSofft — APPROVED** [review]

File: jdi-light-angular-tests/src/main/java/io/github/com/pages/sections/FormFieldsSection.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
+package io.github.com.pages.sections;
+
+import com.epam.jdi.light.angular.elements.complex.*;
```
Comment (gbuh) [inline jdi-light-angular-tests/src/main/java/io/github/com/pages/sections/FormFieldsSection.java:3]: Importing asterisks does not provide much visibility into which types you are using in your code.
This is inconvenient to understand the codebase.

File: jdi-light-angular-tests/src/test/java/io/github/epam/angular/tests/elements/complex/FormFieldTests.java [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+
+    @Test
+    public void formFieldsThemesTest() {
+        formFieldsSection.formFieldThemes.show();
+        formFieldsSection.formFieldThemes.set(1, "Accent");
+        formFieldsSection.formFieldThemes.has().color(1, "rgba(0, 0, 0, 0.87)");
+        formFieldsSection.formFieldThemes.set(2, "20");
+        formFieldsSection.formFieldThemes.has().font(2, containsString("20px"));
+    }
+
```
Comment (gbuh) [inline jdi-light-angular-tests/src/test/java/io/github/epam/angular/tests/elements/complex/FormFieldTests.java:121]: Please remove the empty line

File: jdi-light-angular-tests/src/test/java/io/github/epam/angular/tests/elements/complex/FormFieldTests.java [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+import io.github.epam.*;
+import org.testng.annotations.*;
+
+import static io.github.com.StaticSite.*;
+import static io.github.com.pages.AngularPage.*;
+import static io.github.epam.site.steps.States.*;
+import static org.hamcrest.Matchers.*;
+
+public class FormFieldTests extends TestsInit {
+
```
Comment (gbuh) [inline jdi-light-angular-tests/src/test/java/io/github/epam/angular/tests/elements/complex/FormFieldTests.java:12]: Please remove the empty line

File: jdi-light-angular/src/main/java/com/epam/jdi/light/angular/asserts/FormFieldsAssert.java [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+        jdiAssert(element.label(index), Matchers.is(expected));
+        return this;
+    }
+
+    @JDIAction("Assert that {name} with index {0} has error {1}")
+    public FormFieldsAssert error(int index, String expected) {
+        jdiAssert(element.error(index), Matchers.is(expected));
+        return this;
+    }
+
```
Comment (gbuh) [inline jdi-light-angular/src/main/java/com/epam/jdi/light/angular/asserts/FormFieldsAssert.java:149]: Please remove the empty line

File: jdi-light-angular/src/main/java/com/epam/jdi/light/angular/elements/complex/FormField.java [hunk 11]
```diff
@@ -0,0 +1,322 @@
...
+/**
+ To see an example of FormField web element please visit https://material.angular.io/components/form-field/overview#form-field-appearance-variants
+ **/
+
+public class FormField extends UIBaseElement<FormFieldsAssert> {
+
+    private String matFormFieldPrefix = "//mat-form-field[@";
+    private String autocompleteAttrPrefix = "_ngcontent-";
+
+
```
Comment (gbuh) [inline jdi-light-angular/src/main/java/com/epam/jdi/light/angular/elements/complex/FormField.java:27]: Please remove the empty line

File: jdi-light-angular/src/main/java/com/epam/jdi/light/angular/elements/complex/FormField.java [hunk 12]
```diff
@@ -0,0 +1,322 @@
...
+import static com.epam.jdi.light.common.TextTypes.*;
+import static com.epam.jdi.light.driver.WebDriverFactory.*;
+
+/**
+ To see an example of FormField web element please visit https://material.angular.io/components/form-field/overview#form-field-appearance-variants
+ **/
+
+public class FormField extends UIBaseElement<FormFieldsAssert> {
+
```
Comment (gbuh) [inline jdi-light-angular/src/main/java/com/epam/jdi/light/angular/elements/complex/FormField.java:23]: Please remove the empty line

File: jdi-light-angular/src/main/java/com/epam/jdi/light/angular/elements/complex/FormField.java [hunk 13]
```diff
@@ -0,0 +1,322 @@
...
+        attributes.forEach(attr -> {
+            boolean doesContain = attr.contains(autocompleteAttrPrefix);
+            if (doesContain) {
+                curAttr[0] = attr;
+            }
+        });
+        return curAttr[0];
+    }
+
+    public UIElement getContainer() {
```
Comment (gbuh) [inline jdi-light-angular/src/main/java/com/epam/jdi/light/angular/elements/complex/FormField.java:43]: This method is only used in the current class. Maybe the modifier should be protected.

File: jdi-light-angular/src/main/java/com/epam/jdi/light/angular/elements/complex/FormField.java [hunk 14]
```diff
@@ -0,0 +1,322 @@
...
+
+    @JDIAction("Get hint for input in '{name}' with index {0}")
+    public String inputHint(int index) {
+        UIElement hint = getInputs().get(index).
+            find(By.xpath("//ancestor::mat-form-field[@" + containerAttribute + "]//mat-hint"));
+        return hint.getText();
+    }
+
+    @JDIAction("Get hint for text area in '{name}' with index {0}")
+    public String textAreaHint(int index) {
```
Comment (gbuh) [inline jdi-light-angular/src/main/java/com/epam/jdi/light/angular/elements/complex/FormField.java:148]: This method is never used?

File: jdi-light-angular/src/main/java/com/epam/jdi/light/angular/elements/complex/FormField.java [hunk 15]
```diff
@@ -0,0 +1,322 @@
...
+    }
+
+    @JDIAction("Clear value from text area in '{name}' with index {0}")
+    public void clearTextArea(int index) {
+        getTextAreas().get(index).sendKeys(Keys.CONTROL + "a");
+        getTextAreas().get(index).sendKeys(Keys.DELETE);
+        focusOut();
+    }
+
+    public String getType(int index) {
```
Comment (gbuh) [inline jdi-light-angular/src/main/java/com/epam/jdi/light/angular/elements/complex/FormField.java:207]: This method is only used in the current class. Maybe the modifier should be protected.

File: jdi-light-angular/src/main/java/com/epam/jdi/light/angular/elements/complex/FormField.java [hunk 16]
```diff
@@ -0,0 +1,322 @@
...
+                } catch (Exception e3) {
+                    JDILogger logger = new JDILogger();
+                    logger.error(e3.toString());
+                }
+            }
+        }
+        return type;
+    }
+
+    public WebList getFormFields() {
```
Comment (gbuh) [inline jdi-light-angular/src/main/java/com/epam/jdi/light/angular/elements/complex/FormField.java:231]: This method is only used in the current class. Maybe the modifier should be protected.

File: jdi-light-angular/src/main/java/com/epam/jdi/light/angular/elements/complex/FormField.java [hunk 17]
```diff
@@ -0,0 +1,322 @@
...
+    public String color(int index) {
+        String type = getType(index);
+        WebElement element = getFormFields().get(index).find(By.xpath("//" + type));
+        UIElement actualElement = new UIElement(element);
+        return actualElement.getCssValue("color");
+    }
+
+    @Override
+    public FormFieldsAssert is() { return new FormFieldsAssert().set(this); }
+
```
Comment (gbuh) [inline jdi-light-angular/src/main/java/com/epam/jdi/light/angular/elements/complex/FormField.java:321]: Please remove the empty line

File: jdi-light-angular/src/main/java/com/epam/jdi/light/angular/elements/complex/FormField.java [hunk 18]
```diff
@@ -0,0 +1,322 @@
...
+    @JDIAction("Get color attribute in '{name}' with index {0}")
+    public String color(int index) {
+        String type = getType(index);
+        WebElement element = getFormFields().get(index).find(By.xpath("//" + type));
+        UIElement actualElement = new UIElement(element);
+        return actualElement.getCssValue("color");
+    }
+
+    @Override
+    public FormFieldsAssert is() { return new FormFieldsAssert().set(this); }
```
Comment (gbuh) [inline jdi-light-angular/src/main/java/com/epam/jdi/light/angular/elements/complex/FormField.java:320]: Code convention is not respected. Please format this method.

**Comment by mf-code** [comment]
[6135] - 4W6ixgmqXpZ4

**Comment by mf-code** [comment]
[6135] - uUaYsKkBDyzU

**Comment by mf-code** [comment]
[6135] - ntuDHOKSlaCz

**Comment by mf-code** [comment]
[6135] - xEoF2EaQRjcJ

**Comment by mf-code** [comment]
[6135] - 3Hcv0wDOdw8n

**Comment by mf-code** [comment]
[6135] - im1RWU2X6lAb

File: jdi-light-angular/src/main/java/com/epam/jdi/light/angular/elements/complex/FormField.java (hunk 14 above)
Comment (rralys) [reply jdi-light-angular/src/main/java/com/epam/jdi/light/angular/elements/complex/FormField.java]: Deleted this one.

File: jdi-light-angular/src/main/java/com/epam/jdi/light/angular/elements/complex/FormField.java (hunk 18 above)
Comment (rralys) [reply jdi-light-angular/src/main/java/com/epam/jdi/light/angular/elements/complex/FormField.java]: Updated.

File: jdi-light-angular/src/main/java/com/epam/jdi/light/angular/elements/complex/FormField.java (hunk 17 above)
Comment (rralys) [reply jdi-light-angular/src/main/java/com/epam/jdi/light/angular/elements/complex/FormField.java]: Done.

File: jdi-light-angular/src/main/java/com/epam/jdi/light/angular/elements/complex/FormField.java (hunk 16 above)
Comment (rralys) [reply jdi-light-angular/src/main/java/com/epam/jdi/light/angular/elements/complex/FormField.java]: Done.

File: jdi-light-angular/src/main/java/com/epam/jdi/light/angular/elements/complex/FormField.java (hunk 15 above)
Comment (rralys) [reply jdi-light-angular/src/main/java/com/epam/jdi/light/angular/elements/complex/FormField.java]: Done.

File: jdi-light-angular/src/main/java/com/epam/jdi/light/angular/elements/complex/FormField.java (hunk 13 above)
Comment (rralys) [reply jdi-light-angular/src/main/java/com/epam/jdi/light/angular/elements/complex/FormField.java]: Done.

File: jdi-light-angular/src/main/java/com/epam/jdi/light/angular/elements/complex/FormField.java (hunk 12 above)
Comment (rralys) [reply jdi-light-angular/src/main/java/com/epam/jdi/light/angular/elements/complex/FormField.java]: Done.

File: jdi-light-angular/src/main/java/com/epam/jdi/light/angular/elements/complex/FormField.java (hunk 11 above)
Comment (rralys) [reply jdi-light-angular/src/main/java/com/epam/jdi/light/angular/elements/complex/FormField.java]: Done.

File: jdi-light-angular/src/main/java/com/epam/jdi/light/angular/asserts/FormFieldsAssert.java (hunk 10 above)
Comment (rralys) [reply jdi-light-angular/src/main/java/com/epam/jdi/light/angular/asserts/FormFieldsAssert.java]: Done.

File: jdi-light-angular-tests/src/test/java/io/github/epam/angular/tests/elements/complex/FormFieldTests.java (hunk 9 above)
Comment (rralys) [reply jdi-light-angular-tests/src/test/java/io/github/epam/angular/tests/elements/complex/FormFieldTests.java]: Done.

File: jdi-light-angular-tests/src/test/java/io/github/epam/angular/tests/elements/complex/FormFieldTests.java (hunk 8 above)
Comment (rralys) [reply jdi-light-angular-tests/src/test/java/io/github/epam/angular/tests/elements/complex/FormFieldTests.java]: Done.

File: jdi-light-angular-tests/src/main/java/io/github/com/pages/sections/FormFieldsSection.java (hunk 7 above)
Comment (rralys) [reply jdi-light-angular-tests/src/main/java/io/github/com/pages/sections/FormFieldsSection.java]: Updated.

File: jdi-light-angular-tests/src/main/java/io/github/com/pages/sections/FormFieldsSection.java (hunk 5 above)
Comment (rralys) [reply jdi-light-angular-tests/src/main/java/io/github/com/pages/sections/FormFieldsSection.java]: Changed the variables to use smart locators.

**Comment by mf-code** [comment]
[6143] - 7AFVZK9B2SoO

**Comment by mf-code** [comment]
[6143] - IPlDBR0DiR7O

**Comment by mf-code** [comment]
[6143] - 4N9swLhHsQDx

**Comment by mf-code** [comment]
[6143] - nGYLpMgnrHFX

**Comment by mf-code** [comment]
[6143] - WyoZSaq5pXnM

**Comment by mf-code** [comment]
[6143] - Df7SlkBIQ4gN

**Comment by mf-code** [comment]
[6143] - DCCJ9sP3L9pf

**Comment by mf-code** [comment]
[6143] - rQfQ7wMBk6Tk

**Comment by mf-code** [comment]
[6143] - kwueiOjYpEBx

**Comment by mf-code** [comment]
[6143] - iUbqMi1RQDy2

**Comment by mf-code** [comment]
[6143] - PFPWvd9bHkO4

**Comment by mf-code** [comment]
[6143] - jmuBRnujfQAo

**Comment by mf-code** [comment]
[6143] - tfVtfoVtdCIX

**Comment by mf-code** [comment]
[6143] - 8U753EX6l1VW

**Comment by mf-code** [comment]
[6143] - ATLCKtaIsuiF

**Comment by mf-code** [comment]
[6143] - ZRHuXMA656o4

**Comment by mf-code** [comment]
[6143] - 91VeEiN9Wq5g

**Comment by mf-code** [comment]
[6144] - I5sCsAIlVdpl

**Comment by mf-code** [comment]
[6144] - 8S8wwrExMeXV

**Comment by mf-code** [comment]
[6143] - slyADGsrzUBx

**Comment by mf-code** [comment]
[6144] - DokNJ3a1beyF

**Comment by mf-code** [comment]
[6144] - HnPBkh0DfmaZ

**Comment by mf-code** [comment]
[6144] - c2yPnPJXh3t1

**Comment by mf-code** [comment]
[6144] - 6toqbMg6AeyR

**Comment by mf-code** [comment]
[6144] - sDDUfelEVSyB

**Comment by mf-code** [comment]
[6144] - 4PCPW1NkxQGD

**Comment by mf-code** [comment]
[6146] - rWQbV6zR214n

**Comment by mf-code** [comment]
[6146] - UgU4IWySpeHs

**Comment by mf-code** [comment]
[6146] - IhID1BFWn3MC

**Comment by mf-code** [comment]
[6146] - OcL6QKNR4XUV

**Comment by mf-code** [comment]
[6146] - wnROU9qRcktC

**Comment by mf-code** [comment]
[6146] - fv0fQYI8IF9Z

**Comment by mf-code** [comment]
[6146] - d4Vh6YzZ4MA8

**Comment by mf-code** [comment]
[6146] - c4zlaeN723ka

**Comment by mf-code** [comment]
[6146] - iCT5C5s3UjuL

**Comment by mf-code** [comment]
[6146] - Fh48oUye9SZU

**Comment by mf-code** [comment]
[6146] - BqXeuQFSJa8Z

**Comment by mf-code** [comment]
[6146] - mGYtZvhXWQ6E

**Comment by mf-code** [comment]
[6146] - i6nJCR6BT7hI

**Comment by mf-code** [comment]
[6146] - pip4dzWXensk

**Comment by mf-code** [comment]
[6146] - CdbzsDOMLRog

**Comment by mf-code** [comment]
[6146] - bVy4wyQKfAFN

**Comment by mf-code** [comment]
[6146] - zjOwjT7HcEap

**Comment by mf-code** [comment]
[6146] - YZh4FTlDEfiq

**Comment by mf-code** [comment]
[6146] - WWlUPaJxYWkH

**Comment by mf-code** [comment]
[6146] - mXx9JtiW1EXr

**Comment by mf-code** [comment]
[6146] - dLbrKQpomdqZ

**Pull request merged**